server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -68,8 +68,14 @@ void sendRequest(final DiscoveryNode node, final TcpChannel channel, final long
         Version version = Version.min(this.version, channelVersion);
         OutboundMessage.Request message =
             new OutboundMessage.Request(threadPool.getThreadContext(), request, version, action, requestId, isHandshake, compressRequest);
-        ActionListener<Void> listener = ActionListener.wrap(() ->
-            messageListener.onRequestSent(node, requestId, action, request, options));
+        request.incRef();
```

Comment:
Hmm, I think this bug is really in the recovery code tho, we're completing the overall listener before all the messages have gone, which is kinda lazy. It's ok in this specific case because we don't hold any other resources, but in general I worry that this change might hide different use-after-release bugs.

If this call to `incRef()` fails then IMO the caller is at fault, we should fail tests that hit this. Obtaining another ref to be safe here (and throwing an `AlreadyClosedException` if not) seems like a good backup protection in production tho, but I think we should actually be deliberately lazy in the recovery code so we don't obscure any other use-after-release bugs:

```diff
diff --git a/server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java b/server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java
index db3147c52c2..a1f89da448a 100644
--- a/server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java
+++ b/server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java
@@ -10,6 +10,7 @@ package org.elasticsearch.indices.recovery;

 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.AlreadyClosedException;
 import org.apache.lucene.store.RateLimiter;
 import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.ExceptionsHelper;
@@ -243,8 +244,19 @@ public class RemoteRecoveryTargetHandler implements RecoveryTargetHandler {

             @Override
             public void tryAction(ActionListener<T> listener) {
-                transportService.sendRequest(targetNode, action, request, options,
-                    new ActionListenerResponseHandler<>(listener, reader, ThreadPool.Names.GENERIC));
+                if (request.tryIncRef()) {
+                    transportService.sendRequest(
+                        targetNode,
+                        action,
+                        request,
+                        options,
+                        new ActionListenerResponseHandler<>(
+                            ActionListener.runBefore(listener, request::decRef),
+                            reader,
+                            ThreadPool.Names.GENERIC));
+                } else {
+                    listener.onFailure(new AlreadyClosedException("already closed"));
+                }
             }

             @Override
```

Reply:
++ this makes sense, adjusted the recovery code with the above and made the network side trip an assertion if the request has been released already now.

Reply:
ah no this won't unfortunately, retries will break with this change. Hold off here, I'll see what I can do about this

Reply:
nevermind urgh https://github.com/elastic/elasticsearch/pull/74256/commits/9c51a0c2260fb7a34fbf4851d623a02d0718658d ...  :)